### ISSUE TITLE
std.mem.zeroInit crash zig compiler when used to zero extern struct 

### ISSUE THREAD

**Issue by xnhp0320** [ziglang/zig#16153]
### Zig Version

0.11.0-dev.3771+128fd7dd0

### Steps to Reproduce and Observed Behavior

```zig
   const std = @import("std");

   fn getKey(comptime K:type) type {
       return extern struct{
           key: K,
       };
   }

   fn getSomeType() type {
       var x = std.mem.zeroInit(getKey(u64), .{});
       _ = x;
   }

   test {
       const Y = getSomeType();
       _ = Y;
   }
```

### Expected Behavior

not crash, or throw some errors.

**Comment by xnhp0320** [comment]
What's more weird is that, just change the return type of getSomeType() into void, it becomes normal:
```zig
 const std = @import("std");

 fn getKey(comptime K: type) type {
     return extern struct {
         key: K,
     };
 }

 fn getSomeType() void {
     var x = std.mem.zeroInit(getKey(u64), .{});
     _ = x;
 }

 test {
     const Y = getSomeType();
     _ = Y;
 }

```

**Comment by rohlem** [comment]
>  change the return type of getSomeType() into void, it becomes normal

A function with return type `type` has to be evaluated at `comptime`.
That suggests that the issue is particularly about `std.mem.zeroInit` being used to initialize a `var` at `comptime`.
(I use an older version `0.11.0-dev.3312+ab37ab33c` that doesn't crash on any of these example though.)

**Comment by adimineman** [comment]
Isn't this a duplicate of #16130 ?

**Comment by adimineman** [comment]
Appears to be fixed in master
`user@pc $ zig_nightly test src/test.zig 
src/test.zig:9:18: error: function with non-void return type 'type' implicitly returns
fn getSomeType() type {
                 ^~~~
src/test.zig:12:1: note: control flow reaches end of body here
}
^
src/test.zig:15:26: note: called from here
    const Y = getSomeType();
              ~~~~~~~~~~~^~
`

**Comment by xnhp0320** [comment]
I check with nightly tar ball 0.11.0-dev.3773+438d68091, it still crashed.
do I need to wait for the latest incoming nightly ?

**Comment by rohlem** [comment]
Ah, indeed! @xnhp0320 your commit 438d68091is 6 commits before commit c60896743d3b0776a44ac63582f51b6e64892528, which fixed #16130 .
Since then the macOS-CI seems to have unrelated connection issues; whenever the next master commit gets a green check mark the next ziglang build should have this fixed.

**Comment by xnhp0320** [comment]
@rohlem Thanks for the reply! I will close this issue.
